Microsoft.Toolkit.HighPerformance/Buffers/StringPool.cs

```diff
@@ -0,0 +1,199 @@
+        /// <summary>
+        /// Gets the default <see cref="StringPool"/> instance.
+        /// </summary>
+        public static StringPool Default { get; } = new StringPool();
+
+        /// <summary>
+        /// Gets the total number of buckets in use.
+        /// </summary>
+        public int NumberOfBuckets { get; }
+
+        /// <summary>
+        /// Gets the maximum number of <see cref="string"/> entries stored in each bucket.
+        /// </summary>
+        public int EntriesPerBucket { get; }
+
+        /// <summary>
+        /// Gets a cached <see cref="string"/> instance matching the input content, or creates a new one.
+        /// </summary>
+        /// <param name="span">The input <see cref="ReadOnlySpan{T}"/> with the contents to use.</param>
+        /// <returns>A <see cref="string"/> instance with the contents of <paramref name="span"/>, cached if possible.</returns>
+        public string GetOrAdd(ReadOnlySpan<char> span)
+        {
+            if (span.IsEmpty)
+            {
+                return string.Empty;
+            }
+
+            int bucketIndex = span.Length % NumberOfBuckets;
+
+            Bucket bucket = this.buckets.DangerousGetReferenceAt(bucketIndex);
+
+            lock (bucket)
+            {
+                return bucket.GetOrAdd(span, EntriesPerBucket);
+            }
+        }
+
+        /// <summary>
+        /// Resets the current instance and its associated buckets.
+        /// </summary>
+        public void Reset()
+        {
+            foreach (Bucket bucket in this.buckets)
+            {
+                lock (bucket)
+                {
+                    bucket.Reset();
+                }
+            }
+        }
+
+        /// <summary>
+        /// A configurable bucket containing a group of cached <see cref="string"/> instances.
+        /// </summary>
+        private sealed class Bucket
+        {
+            /// <summary>
+            /// A bitmask used to avoid branches when computing the absolute value of an <see cref="int"/>.
+            /// This will ignore overflows, as we just need this to map hashcodes into the valid bucket range.
+            /// </summary>
+            private const int SignMask = ~(1 << 31);
+
+            /// <summary>
+            /// The array of entries currently in use.
+            /// </summary>
+            private string?[]? entries;
+
+            /// <summary>
+            /// Implements <see cref="StringPool.GetOrAdd"/> for the current <see cref="Bucket"/> instance.
+            /// </summary>
+            /// <param name="span">The input <see cref="ReadOnlySpan{T}"/> with the contents to use.</param>
+            /// <param name="entriesPerBucket">The number of entries being used in the current instance.</param>
+            /// <returns>A <see cref="string"/> instance with the contents of <paramref name="span"/>, cached if possible.</returns>
+            public string GetOrAdd(ReadOnlySpan<char> span, int entriesPerBucket)
```

Comment:
There should be a `TryGet(ROS<char)` method that gets the interned value if is interned, but doesn't if it isn't, similar to `string.IsInterned`

Reply:
Added a new `Add(string)` method in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3380/commits/194e082f118db738af60fe16966dd1472b071dbe, thank you for the suggestion! 🚀